src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -94,100 +94,91 @@ public virtual void GetPolicies(JsonSerializerOptions options)
 
         private void DeterminePropertyName(JsonSerializerOptions options)
         {
-            if (PropertyInfo != null)
+            if (PropertyInfo == null)
             {
-                JsonPropertyNameAttribute nameAttribute = GetAttribute<JsonPropertyNameAttribute>();
-                if (nameAttribute != null)
-                {
-                    NameAsString = nameAttribute.Name;
+                return;
+            }
 
-                    // null is not valid; JsonClassInfo throws an InvalidOperationException after this return.
-                    if (NameAsString == null)
-                    {
-                        return;
-                    }
-                }
-                else if (options.PropertyNamingPolicy != null)
-                {
-                    NameAsString = options.PropertyNamingPolicy.ConvertName(PropertyInfo.Name);
+            JsonPropertyNameAttribute nameAttribute = GetAttribute<JsonPropertyNameAttribute>(PropertyInfo);
+            if (nameAttribute != null)
+            {
+                NameAsString = nameAttribute.Name;
 
-                    // null is not valid; JsonClassInfo throws an InvalidOperationException after this return.
-                    if (NameAsString == null)
-                    {
-                        return;
-                    }
-                }
-                else
+                if (NameAsString == null)
                 {
-                    NameAsString = PropertyInfo.Name;
+                    ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameNull(ParentClassType, this);
                 }
+            }
+            else if (options.PropertyNamingPolicy != null)
+            {
+                NameAsString = options.PropertyNamingPolicy.ConvertName(PropertyInfo.Name);
 
-                // At this point propertyName is valid UTF16, so just call the simple UTF16->UTF8 encoder.
-                _name = Encoding.UTF8.GetBytes(NameAsString);
-
-                // Set the compare name.
-                if (options.PropertyNameCaseInsensitive)
+                if (NameAsString == null)
                 {
-                    NameUsedToCompareAsString = NameAsString.ToUpperInvariant();
-                    _nameUsedToCompare = Encoding.UTF8.GetBytes(NameUsedToCompareAsString);
-                }
-                else
-                {
-                    NameUsedToCompareAsString = NameAsString;
-                    _nameUsedToCompare = _name;
+                    ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameNull(ParentClassType, this);
                 }
+            }
+            else
+            {
+                NameAsString = PropertyInfo.Name;
+            }
+
+            // At this point propertyName is valid UTF16, so just call the simple UTF16->UTF8 encoder.
+            _name = Encoding.UTF8.GetBytes(NameAsString);
```

Comment:
From https://github.com/dotnet/corefx/pull/37578#discussion_r283076379

> Also, is `_name` a private property here rather than a field? Why? Maybe using PascalCasing or make it a field?

> `_name` is a private field.

`_name` has a getter/setter. It looks like its a property to me: https://github.com/dotnet/corefx/blob/33247077de38f44d68d18e5b3bd11fe341f28bd0/src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs#L28

If the intention is for it to be a field, I am fine with that, but we should remove the getter/setter.

Reply:
You are correct. It is a property. At one point it was a field. This data is on the "hot path" of the stack-based state so I was using fields which are faster than properties, so there is\was a mix of field access and some protected property setter for cases that aren't on the "hot path". However I find that doing this is causing more trouble in reviews than it is worth so I will make them all properties (if accessed outside the class) and measure perf impact at a later stage when we micro-optimize.

Reply:
I have no concerns with it being fields (especially if you previously observed perf difference). I have just gotten used to properties being pascal cased, so having a "field-like" name threw me off.